### PR TITLE
RavenDB-19149 Lucene: Optimize method 

### PR DESCRIPTION
**Pull request by maciejaszyk**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-19149 

### Additional description

This PR contains two things:
- When Optimize is running we're turning off timeout in `TimeTrackingSerialMergeScheduler` because we cannot move any merge into the next batch. Everything has to be done in one transaction.
- Abort - `CancellationToken` support in `SegmentsMerger` when `Optimize` is called.

### Type of change

- Bug fix
- New feature

### How risky is the change?

- Moderate 

### Backward compatibility

- Non breaking change

### Is it platform specific issue?

- No

### Documentation update

- No documentation update is needed 

### Testing 

- Tests have been added that prove the fix is effective or that the feature works

### Is there any existing behavior change of other features due to this change?

- No

### UI work

- Yes

### CODE REVIEW

File: test/FastTests/Issues/RavenDB-6250.cs [hunk 1]
```diff
@@ -55,7 +55,8 @@ public void All_operations_has_details_providers()
             var operationWithoutDetails = new HashSet<Operations.OperationType>
             {
                 Operations.OperationType.Setup, // it is in secured setup,
-                Operations.OperationType.DebugPackage
+                Operations.OperationType.DebugPackage,
+                Operations.OperationType.LuceneOptimizeIndex
```
Comment (maciejaszyk) [inline test/FastTests/Issues/RavenDB-6250.cs:59]: @ml054 should I add something in the studio too?

File: src/Raven.Client/ServerWide/CompactSettings.cs [hunk 2]
```diff
@@ -9,14 +9,17 @@ public class CompactSettings
         public bool Documents { get; set; }
 
         public string[] Indexes { get; set; }
+        
+        public bool SkipOptimize { get; set; }
```
Comment (arekpalinski) [inline src/Raven.Client/ServerWide/CompactSettings.cs]: We need to be more specific and explicitly say that it's about indexes so maybe `SkipOptimizeIndexes`.

This option should be exposed in the studio CC @ml054

File: src/Raven.Server/Documents/Handlers/IndexHandler.cs [hunk 3]
```diff
@@ -1002,6 +1005,54 @@ await using (var ms = new MemoryStream())
...
+            using (ServerStore.ContextPool.AllocateOperationContext(out TransactionOperationContext context))
+            {
+                var name = GetQueryStringValueAndAssertIfSingleAndNotEmpty("name");
+                var index = Database.IndexStore.GetIndex(name);
+                if (index == null)
+                    throw new InvalidDataException($"Index '{name}' not found. Cannot perform optimize.");
+                var token = CreateOperationToken();
+                using (index.DrainRunningQueries())
+                {
+                    var result = new CompactionResult(index.Name);
```
Comment (arekpalinski) [inline src/Raven.Server/Documents/Handlers/IndexHandler.cs]: `CompactionResult`?

File: src/Raven.Server/Documents/Handlers/IndexHandler.cs [hunk 4]
```diff
@@ -1002,6 +1005,54 @@ await using (var ms = new MemoryStream())
...
+                        "Optimizing index: " + index.Name,
+                        Operations.Operations.OperationType.LuceneOptimizeIndex,
+                        taskFactory: onProgress => Task.Run(() =>
+                        {
+                            try
+                            {
+                                using (token)
+                                using (Database.PreventFromUnloadingByIdleOperations())
+                                using (var indexCts = CancellationTokenSource.CreateLinkedTokenSource(token.Token, Database.DatabaseShutdown))
+                                using (index.RestartEnvironment(onBeforeEnvironmentDispose: () => index.Optimize(onProgress, result, indexCts.Token)))
```
Comment (arekpalinski) [inline src/Raven.Server/Documents/Handlers/IndexHandler.cs]: This way we ensure that we are currently running any indexing, right?

Comment (arekpalinski) [reply src/Raven.Server/Documents/Handlers/IndexHandler.cs]: I don't see a problem with restarting the environment (even if not required for this operation), just asking to be sure

File: src/Raven.Server/Documents/Handlers/IndexHandler.cs [hunk 5]
```diff
@@ -1002,6 +1005,54 @@ await using (var ms = new MemoryStream())
...
+                                {
+                                    return Task.FromResult((IOperationResult)result);
+                                }
+                            }
+                            catch (Exception e)
+                            {
+                                if (Logger.IsOperationsEnabled)
+                                    Logger.Operations("Compaction process failed", e);
```
Comment (arekpalinski) [inline src/Raven.Server/Documents/Handlers/IndexHandler.cs]: Compaction -> Optimize

File: src/Raven.Server/Documents/Handlers/IndexHandler.cs (hunk 4 above)
Comment (maciejaszyk) [reply src/Raven.Server/Documents/Handlers/IndexHandler.cs]: I believe so, this is how we handle `Optimize` in the Compact handler.

File: src/Raven.Server/Documents/Handlers/IndexHandler.cs (hunk 3 above)
Comment (maciejaszyk) [reply src/Raven.Server/Documents/Handlers/IndexHandler.cs]: Left as mock for Operation task

File: src/Raven.Server/Documents/Handlers/IndexHandler.cs (hunk 5 above)
Comment (maciejaszyk) [reply src/Raven.Server/Documents/Handlers/IndexHandler.cs]: done

File: src/Raven.Server/Documents/Indexes/Index.cs [hunk 6]
```diff
@@ -4698,25 +4703,46 @@ public void Compact(Action<IOperationProgress> onProgress, CompactionResult resu
...
+        internal void Optimize(Action<IOperationProgress> onProgress, CompactionResult result, CancellationToken token)
+        {
+            if (_isIndexOptimizationInProgress)
+                throw new InvalidOperationException($"Index '{Name}' cannot be optimized because optimization is already in progress.");
+            
+            try
+            {
+                _isIndexOptimizationInProgress = true;
+                result.AddMessage($"Starting data optimization of index '{Name}'.");
+                onProgress?.Invoke(result.Progress);
```
Comment (arekpalinski) [inline src/Raven.Server/Documents/Indexes/Index.cs]: Couldn't the above 2 lines be called inside `Compact()` method? This way so won't need so pass `CompactionResult` from the endpoint

https://github.com/ravendb/ravendb/pull/15136/files#diff-897d9d741374625268d6ede98676854223bfe67e16811b5f468babe94de6d9feR1020

File: src/Raven.Server/Documents/Handlers/IndexHandler.cs (hunk 3 above)
Comment (arekpalinski) [reply src/Raven.Server/Documents/Handlers/IndexHandler.cs]: Check my below comment:

https://github.com/ravendb/ravendb/pull/15136/files#r1001419051

I think we should have some IndexOptimizeResult here

Comment (maciejaszyk) [reply src/Raven.Server/Documents/Handlers/IndexHandler.cs]: Done

File: src/Raven.Server/Documents/Indexes/Index.cs (hunk 6 above)
Comment (maciejaszyk) [reply src/Raven.Server/Documents/Indexes/Index.cs]: moved

**Review by arekpalinski — APPROVED** [review]

**Comment by arekpalinski** [comment]
please fix studio compilation

**Review by arekpalinski — APPROVED** [review]

File: src/Raven.Server/Documents/Handlers/IndexHandler.cs [hunk 7]
```diff
@@ -1002,6 +1005,58 @@ await using (var ms = new MemoryStream())
             }
         }
 
+        [RavenAction("/databases/*/indexes/optimize", "POST", AuthorizationStatus.ValidUser, EndpointType.Write, DisableOnCpuCreditsExhaustion = true)]
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Handlers/IndexHandler.cs]: AuthorizationStatus.DatabaseAdmin and move to AdminIndexHandler `/databases/*/admin/indexes/optimize`

**Review by ppekrol — COMMENTED** [review]
Some of the feedback was done face-to-face

**Pull request merged**